### PR TITLE
Android: make sure the activity gets really finished

### PR DESCRIPTION
**Pull request by 1bsyl**
https://github.com/libsdl-org/SDL/issues/7567

### CODE REVIEW

File: android-project/app/src/main/java/org/libsdl/app/SDLActivity.java [hunk 1]
```diff
@@ -324,6 +325,15 @@ protected void onCreate(Bundle savedInstanceState) {
         Log.v(TAG, "onCreate()");
         super.onCreate(savedInstanceState);
 
+
+        /* Really finish the activity, even if user try to re-start it quickly */
+        if (mIsFinishing != 0 && mIsFinishing < 5) {
+            Log.v(TAG, "finish activity");
+            SDLActivity.mIsFinishing += 1;
+            finish();
```
Comment (slouken) [inline android-project/app/src/main/java/org/libsdl/app/SDLActivity.java:333]: Do we need to call finish() again?

Comment (1bsyl) [reply android-project/app/src/main/java/org/libsdl/app/SDLActivity.java]: indeed, it's not needed ! just updated!

**Comment by slouken** [comment]
FYI, it is entirely legitimate in the Android life cycle for the application to stay resident and then get onCreate() multiple times. I don't think we can assume that it will be unloaded, we just have to warn people that it can happen and they should clean up statics before exit.

**Comment by 1bsyl** [comment]
> FYI, it is entirely legitimate in the Android life cycle for the application to stay resident and then get onCreate() multiple times. I don't think we can assume that it will be unloaded, we just have to warn people that it can happen and they should clean up statics before exit.

yes, but here, the finish() command initially comes after the end of the user main(). it's somehow aborted when user quickly re-create. we shouldn't let the user recreate the activity, because it re-use the same static data (sdl and also app).

**Comment by 1bsyl** [comment]
waiting testing from https://github.com/libsdl-org/SDL/issues/7567

**Comment by slouken** [comment]
However, what can happen now is that the application remains resident and the user will try to relaunch 5 times before it finally succeeds (based on your test), reusing the existing static data. What we should actually do is fix the bug causing the ANR and allow SDL to be relaunched with the previous run static data.

**Comment by 1bsyl** [comment]
The explicit call to finish() (or is simply "return" really equivalent ?):
https://developer.android.com/reference/android/app/Activity#finish()
`Call this when your activity is done and should be closed. The ActivityResult is propagated back to whoever launched you via onActivityResult().`
we should really terminate the activity and all stuff when "main()" ends

I don't think re-using the static data should be ok. all apps have the bad pratice of have somewhere a static global variable that will crash or make a bug when, re-used. And it's not something people are user to testing.

**Comment by 1bsyl** [comment]
( but we  sequence " Init() Quit() Init() Quit()"  should be working if possible, as long as it's in the main())

**Comment by madebr** [comment]
I'm not an Android specialist, but can we use [`finish()`](https://developer.android.com/reference/android/app/Activity#finish()) and [`isFinishing()`](https://developer.android.com/reference/android/app/Activity#isFinishing())?

**Comment by 1bsyl** [comment]
@madebr 
Yes, we use finish and isFinishing() in normal case, but here's that's a corner-case. the previous finish() get aborted: user re-click (eg onCreate), before the onDestroy can be called.

what happens in one of the normal case:
1/ the c main() returns. (it's in a java thread)
2/ we trigger the Activity finish()
 https://github.com/libsdl-org/SDL/blob/main/android-project/app/src/main/java/org/libsdl/app/SDLActivity.java#L1905
3/ then the normal step, is that stop() get called, unfocus(), and onDestroy(). and all is gone. (should be).

the bug - that I can reproduce - is:
1/ same
2/ same
3/ user click quickly again the app. before on Destroy is called. so onDestroy is not called. and the app just go back to onCreate().
 the libSDL.so is alread loaded and the static data are already initialized : in liSDL.so, but also in the SDLActivity.
 and this is clearly a wrong situation.

so the solution: 
1/ same
2/ same + marked a static var mIsFinishing
3/ if we ever go back to onCreate(), we return and try to finish again (up to arbitrary 5x)


I will revert the first modif, to again explicitly call finish() in onCreate so that it's clear we attempt to finish the activity

**Comment by 1bsyl** [comment]
calling exit() seems better than finish(), 

if i do the quick sequence close-restart, then:
    before (finish()) it was like a bad click: app didn't start again (because the 2nd call was to finish the first one)
    now (exit()), it exits the activity with the message, but recreate a new one just after: it clean up the previous activity and start a new one. so the behavior looks good to me

**Comment by 1bsyl** [comment]
probably the counter `mIsFinishing < 5` is now unneeded

**Comment by slouken** [comment]
Yup. Make sure this behavior is gated on a hint. There are applications that use other UI and create SDL activities on demand, and they should not be shut down when the SDL activity finishes.

**Comment by 1bsyl** [comment]
ok. I've updated the PR + documented it.

File: include/SDL3/SDL_hints.h [hunk 2]
```diff
@@ -145,6 +145,19 @@ extern "C" {
  */
 #define SDL_HINT_ANDROID_TRAP_BACK_BUTTON "SDL_ANDROID_TRAP_BACK_BUTTON"
 
+/**
+ * \brief A variable to control whether SDL activity is allowed to be re-created.
+ *        If so, java static datas and static datas from native libraries remain with their current values.
+ *        When not allowed, the activiy terminates with exit(0) to be fully re-initialized afterward.
```
Comment (slouken) [inline include/SDL3/SDL_hints.h:151]: typo: "activiy"

**Comment by slouken** [comment]
It looks like this logic will end up causing the user to launch the game once, then try to relaunch, cause it to exit, and then have to relaunch again. In addition, the hints are cleared on SDL_Quit(), so the hint may no longer be set if the application shuts down properly.

What we should do instead is register the hint to a callback in C code that sets a global variable that we check when the activity completes, and then exit after finishing if needed.

File: include/SDL3/SDL_hints.h (hunk 2 above)
Comment (1bsyl) [reply include/SDL3/SDL_hints.h]: fixed

**Comment by 1bsyl** [comment]
> In addition, the hints are cleared on SDL_Quit(), so the hint may no longer be set if the application shuts down properly.
> What we should do instead is register the hint to a callback in C code that sets a global variable that we check when the activity completes, and then exit after finishing if needed.

updated, with also an AtomicInt, since the SDLActivity could be reading when the c thread is writing the value.


> It looks like this logic will end up causing the user to launch the game once, then try to relaunch, cause it to exit, and then have to relaunch again.

ok. 
1/ this issue is a corner-case.
In normal case (99,9% case). you just quit the app. and 2seconds later the activity is destroyed, since we call 
`SDLActivity.mSingleton.finish();` after the c thread ends.
https://github.com/1bsyl/SDL/blob/main/android-project/app/src/main/java/org/libsdl/app/SDLActivity.java#L1905

2/  It **looks like** this logic will end up causing the user to launch the game once, then try to relaunch, cause it to exit, and then have to relaunch again.

yes, it **looks like**, but this isn't what's happening in practice.
You don't have to **relaunch again**. it's automatically relaunching. I agree, this may be surprising, but that's what's happening.

Anyway, even if user has to relaunch again. it's better than an ANR or game blocked with a black screen.

**Comment by AntTheAlchemist** [comment]
@1bsyl Tested *just* before your AtomicInt update; so I'll test the update again later. It does seem to work and restarts automatically on all the devices I tried. With the side-effect of resetting global native data, which isn't a bad thing.

The exit() call must be happening to old instance, so the atomic check makes sure there's no rare-case crossover where it exists the new instance? Even if it causes a user to retry to relaunch, yes that's far better than an ANR / block.

I haven't read the entire discussion on this, but I guess waiting for the old thread to naturally quid has been ruled out?

The hint only works if you call it *after* `SDL_Quit()`, because `SDL_Quit()` must be resetting the hints? It does work if set the hint in the manifest though.

**Comment by AntTheAlchemist** [comment]
So the `EGL_BAD_ALLOC` when destroying and trying to recreate a window on Android is a seperate unrelated but? Can it be fixed? Shall I create a new issue for it?

**Comment by 1bsyl** [comment]
> @1bsyl Tested _just_ before your AtomicInt update; so I'll test the update again later. It does seem to work and restarts automatically on all the devices I tried. With the side-effect of resetting global native data, which isn't a bad thing.
> 
> The exit() call must be happening to old instance, so the atomic check makes sure there's no rare-case crossover where it exits the new instance? Even if it causes a user to retry to relaunch, yes that's far better than an ANR / block.
> 
> I haven't read the entire discussion on this, but I guess waiting for the old thread to naturally quid has been ruled out?
> 
> The hint only works if you call it _after_ `SDL_Quit()`, because `SDL_Quit()` must be resetting the hints? It does work if set the hint in the manifest though.

I've tested the current version, and it works with setting the hint via AndroidManifest.xml.
It also works before SDL_Quit since slouken suggest to use a callback for it. (must be fixed with the new version).

Honestly, I don't understand how one can run sdl a second time in a safe way, with the static datas.
it means: 
- at java level: you re-create surfaceview with all it pause/resume state, layout, reset orientation, darkmode, HID, 
- at native level: all global variable: at set, mutex, all drivers were init once


I also still don't understand how you manage to keep the activity running for a while, after the SDL main() ends, and finish() called ? 


I will create a new ticket for second. (btw it's about Init/Quit x2,  destroy/create window should be fine).

**Comment by AntTheAlchemist** [comment]
Is there a reason why this hasn't been merged yet? I'm just about to test this live on Google Play. My ANRs are way above bad behaviour, so I'm hoping this will make a difference. The top offending ANR being reported is "Input dispatching timed out" in ".MainActivity.onDestroy". 77 ANRs reported in the past 28 days, accounting for 30.2% of all ANRs. The rest are caused by SDL_LockMutex - I'm hoping Sam can address these in another issue I posted.

I checked Android Studio's profiler before this patch and it looks like there's some stale processes that pile up each time the app is restarted. After this patch, it seems to fix this, since the activity is now destroyed. 

Looking forward to seeing the difference this makes. My app has 600 new downloads daily, so I should notice the effect. Mostly Android TV devices.

**Pull request merged**